### ISSUE TITLE
Kirin spawns multiple pets at once (potential fixes in comment)

### ISSUE THREAD

**Issue by neosatus** [myera/issues#126]
The intention of the code is that it should spawn one pet once every 180 seconds. This is seen in the if statement condition, if battleTime is not 0 and battleTime % 180 == 0 then pick one random pet that hasn't been spawned yet and spawn it.

In game, we're seeing multiple pets spawn at once. I believe this is happening because the function this resides in, onMobFight in the Kirin.lua script, fires off every processor tick (according to Warp) which is multiple times per second causing the if condition to pass multiple times in a row. Apparently 2-3 times per second given the results we're seeing in game as we never saw all 4 pop, nor only 1, it varied from 2-3.

Adding a sleep(1) call to the end of the if statement should fix it, though I don't know if sleeping will cause other issues... at most it should only cost you a second every 3 minutes and never again once all 4 pets are spawned. Maybe there's a more elegant fix though, such as setting a timeout variable and checking for that in the conditional as well.

**Comment by Darkstorm87** [comment]
Perhaps you can do something like this.  I'm not sure of the syntax at all, but perhaps the concept will work.

I added a 'local var' called kirinAddNum.  On spawn it gets set to 1.  Inside the onMobFight function, we do some division and floor it, to get how many 180s have passes since the start of the fight.  We then compare that to the kirinAddNum to see if we can enter the if statement that controls which add to pop.  Immediately inside there, we increment that local var, so the next attempts to enter the if statement fail until the next increment of 180. 

```
-----------------------------------
-- onMobSpawn Action
-----------------------------------

function onMobSpawn(mob)
    mob:setMod(MOD_WINDRES, -64);
	mob:setLocalVar("kirinAddNum", 1);
end

-----------------------------------
-- onMobFight Action
-----------------------------------

function onMobFight( mob, target )
    if (mob:getHPP() < math.random(50,60) and mob:getLocalVar("astralFlow") == 0) then
        mob:useMobAbility(734);
        -- Spawn Avatar
        mob:spawnPet();
        mob:setLocalVar("astralFlow", 1);
    end
	
	local kirinAddNum = mob:getLocalVar("kirinAddNum")
	
    if (mob:getBattleTime() ~= 0 and mob:getBattleTime() % 180 == 0 and  math.floor(mob:getBattleTime() / 180) == kirinAddNum) then
		mob:setLocalVar("kirinAddNum", kirinAddNum + 1); -- increment the counter so you don't reenter until next 180 seconds
		
        -- Ensure we have not spawned all pets yet..
        local genbu = mob:getLocalVar("genbu");
        local seiryu = mob:getLocalVar("seiryu");
        local byakko = mob:getLocalVar("byakko");
        local suzaku = mob:getLocalVar("suzaku");

        if (genbu == 1 and seiryu == 1 and byakko == 1 and suzaku == 1) then
            return;
        end

        -- Pick a pet to spawn at random..
        local ChosenPet = nil;
        local newVar = nil;
        repeat

            local rand = math.random( 0, 3 );
            ChosenPet = 17506671 + rand;

            switch (ChosenPet): caseof {
                [17506671] = function (x) if ( genbu == 1) then ChosenPet = 0; else newVar = "genbu";  end end, -- Genbu
                [17506672] = function (x) if (seiryu == 1) then ChosenPet = 0; else newVar = "seiryu"; end end, -- Seiryu
                [17506673] = function (x) if (byakko == 1) then ChosenPet = 0; else newVar = "byakko"; end end, -- Byakko
                [17506674] = function (x) if (suzaku == 1) then ChosenPet = 0; else newVar = "suzaku"; end end, -- Suzaku
            }

        until (ChosenPet ~= 0 and ChosenPet ~= nil)

        -- Spawn the pet..
        local pet = SpawnMob( ChosenPet );
        pet:updateEnmity( target );
        pet:setPos( mob:getXPos(), mob:getYPos(), mob:getZPos() );

        -- Update Kirins extra vars..
        mob:setLocalVar(newVar, 1);
    end

    -- Ensure all spawned pets are doing stuff..
    for pets = 17506671, 17506674 do
        if (GetMobAction( pets ) == 16) then
            -- Send pet after current target..
            GetMobByID( pets ):updateEnmity( target );
        end
    end
end

```

**Comment by neosatus** [comment]
I came up with this easier way to do it and tested it with Loco, confirmed it works:

File: /scripts/zones/The_Shrine_of_RuAvitau/Kirin.lua

New line at 26: mob:setLocalVar("numAdds", 1);
New line at 40: local numAdds = mob:getLocalVar("numAdds");
Change line 41: if (mob:getBattleTime() / 180 == numAdds) then
New line 76: mob:setLocalVar("numAdds", numAdds + 1);

**Comment by Darkstorm87** [comment]
Ahh, fair enough.  The logic should work.

I didn't know what kind of overhead the mob:getLocalVar("numAdds") call would do, since it has to go back to core every execution to get that var.  I figured you might as well not go fetch the var unless you absolutely needed to.

Also, I didn't know if anything ran async and incrementing with setLocalVar wouldn't happen before a separate call to the function occurred...potentially letting another iteration through the popping function.  That's why I incremented it immediately, not fool proof, but might help if anything is async.

Cool stuff though.  Can't wait to get after Kirin.  Thanks
